cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,313 @@
+ * If the element is null, it will return `ResultType{identity}`.
+ *
+ * @tparam has_nulls Indicates if the column contains null values 
+ *                   (`null_count > 0`)
+ * @tparam ElementType The cudf data type of input array
+ * @tparam ResultType  cudf data type of output and `identity` value
+ *                     which is used when null bitmaps flag is false.
+ * @tparam Iterator_Index
+ *                     The base iterator which gives the index of array.
+ *                     The default is `thrust::counting_iterator`
+ *
+ * @param[in] column   The column device view of cudf column 
+ * @param[in] identity The identity value used when the mask value is false
+ * @param[in] it       The index iterator, `thrust::counting_iterator` by default
+ * -------------------------------------------------------------------------**/
+template <bool has_nulls, typename ElementType, typename ResultType = ElementType,
+    typename Iterator_Index=thrust::counting_iterator<cudf::size_type> >
+auto make_iterator(const column_device_view& column,
+    ResultType identity=ResultType{0}, Iterator_Index const it = Iterator_Index(0))
+{
+    // check the data type
+    CUDF_EXPECTS(data_type(experimental::type_to_id<ElementType>()) == column.type(), "the data type mismatch");
+
+    CUDF_EXPECTS(column.head() != nullptr, "non-null data is required");
+    CUDF_EXPECTS(not ( has_nulls == true && !column.nullable() ), "non-null bit mask is required");
+
+    return thrust::make_transform_iterator(
+      it, experimental::value_accessor<ElementType, ResultType, has_nulls>{column, identity});
+}
+
+/** -------------------------------------------------------------------------*
+ * @brief Constructs an iterator over the elements of a column
+ * Input iterator which can be used for cub and thrust.
+ * 
+ * The iterator returns thrust::pair<ResultType, bool>
+ * This is useful for more complex logic that depends on the validity.
+ * e.g. group_by.count, mean_var, sort algorism.
+ *
+ * @tparam has_nulls   True if the data has valid bit mask, False else
+ * @tparam ElementType The cudf data type of input array
+ * @tparam ResultType  cudf data type of output and `identity` value
+ *                     which is used when null bitmaps flag is false.
+ * @tparam Iterator_Index
+ *                     The base iterator which gives the index of array.
+ *                     The default is `thrust::counting_iterator`
+ *
+ * @param[in] column   The column device view of cudf column 
+ * @param[in] identity The identity value used when the mask value is false
+ * @param[in] it       The index iterator, `thrust::counting_iterator` by default
+ * -------------------------------------------------------------------------**/
+template <bool has_nulls, typename ElementType, typename ResultType = ElementType,
+    typename Iterator_Index=thrust::counting_iterator<cudf::size_type> >
+auto make_pair_iterator(const column_device_view& column,
```

Comment:
Likewise, this does too many things. 

How about instead:
```
/**
 * @brief Constructs an iterator over the  elements of a column that returns a
 * pair containing the element value and it's validity.
 *
 * Dereferencing the returned iterator for element `i` will return a
 * `thrust::pair<Element, bool>` that holds the value of element `i` and whether
 * or not it is valid. `true` is valid, `false` is null.
 *
 * @tparam Element The type of elements in the column
 * @param column The column to iterate
 * @return auto Iterator that returns a pair holding the element value and
 * whether or not it is valid.
 */
template <typename Element>
auto make_pair_iterator(column_device_view column)
```

Reply:
Likewise, there's no need for the `has_nulls` template parameter.

Reply:
This point is addressed as separate PR https://github.com/rapidsai/cudf/pull/32981